File: hangman.py
import random
import sys
import os


if os.name == "posix":
    CLEAN_CMD = "clear"
elif os.name == "nt":
    CLEAN_CMD = "cls"
else:
    raise RuntimeWarning(
        """Clearing command line for operative systems others than Windows, OS X and Linux distributions is not supported"""
    )


def clean_screen() -> None:
    os.system(CLEAN_CMD)


def get_word(filename: str) -> str:
    with open(filename, 'r') as f:
        words = f.readlines()
    word = random.choice(words).strip().upper()
    return word


def make_guess() -> str:
    char = input("Insert an alphanumerical character (or 1 to quit): ").upper()
    while not (char.isalpha or char == '1'):
        return make_guess()
    return char


def replace_char(char: str, src: str, to: str) -> str:
    to_chars = list(to)
    for i in range(len(src)):
        if src[i] == char:
            to_chars[i] = char
    return "".join(to_chars)


def render_screen(lives: int, word: str, wrong_attempts) -> None:
        print(f"Lives remaining: {lives}")
        print(f"The word is: {word}")
        if wrong_attempts:
            print(f"Your wrong guesses are: ", end='')
            print(*wrong_attempts, sep=', ')


def play(sec_word: str) -> int:
    player_word: str = '_' * len(sec_word)
    wrong_guesses = set()
    lives: int = 5
    while lives > 0 and player_word != sec_word:
        clean_screen()
        render_screen(lives, player_word, wrong_guesses)
        guess = make_guess()
        if guess == '1': break
        elif guess in sec_word:
            player_word = replace_char(guess, sec_word, player_word)
        else:
            if not guess in wrong_guesses:
                lives -= 1
                wrong_guesses.add(guess)

    clean_screen()
    if guess == '1':
        out = -1
    elif lives == 0:
        out = 0
    else:
        out = 1

    return out


def main() -> None:
    if len(sys.argv) != 2:
        print("Usage: python3 hangman.py <file>")
        return
    filename = sys.argv[1]
    secret_word = get_word(filename)
    game_status = play(secret_word)
    if game_status == 1:
        print(f"Congratulations! You won :)")
    elif game_status == -1:
        print(f"You finished the game")
    else:
        print(f"You lost, but keep trying")
    print(f"The word was: << {secret_word} >>")


if __name__ == '__main__':
    main()

File: src/hangman/hangman.c
#include "hangman.h"

void init(char *word, int len, char attempts[LIVES-1]) {
    for (int i = 0; i < len; word[i] = '_', i++);
    word[len] = '\0';
    for (int i = 0; i < LIVES-1; i++)
        attempts[i] = '\0';
}

char make_guess() {
    char c;
    printf("Input an alphabetic character (or 1 to quit): ");
    do {
        scanf("%c", &c);
    } while (!isalpha(c) && c != '1');
    return toupper(c);
}

char replace_char(char c, const char* _from, char* _to) {
    
    for (; *_from && *_to; _from++, _to++) {
        if (*_from == c)
            *_to = c;
    }
    return c;
}

int tried_attempt(char c, char attempts[LIVES-1]) {
    int i;
    for (i = 0; attempts[i] != '\0'; i++) {
        if (attempts[i] == c)
            return 1;
    }
    attempts[i] = c;
    return 0;
}

void render_screen(int current_lives, const char* player_word, char attempts[LIVES-1]) {
    printf("Lives remaining: %d\n", current_lives);
    printf("The word is: %s\n", player_word);
    printf("Your wrong guesses are:");
    while (*attempts != '\0') {
        printf(" %c", *attempts);
        attempts++;
    }
    printf("\n");
}

int play(const char *secret_word)
{
    int current_lives = LIVES;
    const int len = strlen(secret_word);
    char previous_attempts[LIVES - 1];
    char *player_word = malloc(sizeof(char) * (len + 1));
    init(player_word, len, previous_attempts);
    char guess;

    while (strcmp(player_word, secret_word) != 0 && current_lives > 0) {
        clear_screen();
        render_screen(current_lives, player_word, previous_attempts);
        guess = make_guess();
        if (guess == '1') {
            break;
        } else if (strchr(secret_word, guess) != NULL) {
            replace_char(guess, secret_word, player_word);
        } else {
            if (tried_attempt(guess, previous_attempts))
                continue;
            current_lives--;
        }
    }
    clear_screen();
    free(player_word);

    if (guess == '1') {
        return -1;
    }
    return current_lives;
}